apache/zeppelin
# [ZEPPELIN-901] Cassandra interpreter V3

### What is this PR for?
**Cassandra** interpreter V3:

* Update documentation
* Update interactive documentation
* Add support for binary protocol **V4**
* Implement new `@requestTimeOut` runtime option
* Upgrade Java driver version to **3.0.1**
* Allow interpreter to add dynamic forms programmatically when using `FormType.SIMPLE`
* Allow dynamic form using default Zeppelin syntax
* Fixing typo on `FallThroughPolicy`
* Look for data in `AngularObjectRegistry` before creating dynamic form
* Add missing support for `ALTER` statements

### What type of PR is it?
[Improvement]

### Todos
* [ ] - Test steps executed and confirmed working by the community

### What is the Jira issue?
**[ZEPPELIN-901]**

### How should this be tested?
* Download and install locally **[Cassandra 3.5]**
* Start **Cassandra**
* Clone this pull request locally with:
    * `git fetch origin pull/950/head:CassandraInterpreterV3`
    * `git checkout CassandraInterpreterV3`
* Build this version of **Zeppelin** with `mvn clean package -DskipTests`
* Start Zeppelin and update the property `cassandra.hosts` of the **Cassandra** interpreter (set it to _localhost_ or _127.0.0.1_ depending on your configuration)
* To test the `ALTER` statement, create a new paragraph with the following content:
```
%cassandra

CREATE KEYSPACE IF NOT EXISTS cassandra_v3 WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 1};
CREATE TABLE IF NOT EXISTS cassandra_v3.simple(key int PRIMARY KEY);
ALTER TABLE cassandra_v3.simple ADD value text;
DESCRIBE TABLE cassandra_v3.simple;
```
The result of this paragraph should be:
![image](https://cloud.githubusercontent.com/assets/1532977/15740044/6062bba4-28b3-11e6-818e-e257d0830041.png)
 
* To test the support for **Zeppelin** dynamic form syntax, create a paragraph with the following content:

```
%cassandra

INSERT INTO cassandra_v3.simple(key,value) VALUES(1, 'first'); 
INSERT INTO cassandra_v3.simple(key,value) VALUES(2, 'second');
INSERT INTO cassandra_v3.simple(key,value) VALUES(3, 'third');

SELECT * FROM cassandra_v3.simple WHERE key=${key=1};
```
The result of this paragraph should be:
![image](https://cloud.githubusercontent.com/assets/1532977/15740146/d7b696f8-28b3-11e6-80c5-428ae97558f6.png)
* To test the legacy syntax for dynamic form, create a new paragraph with the following content:

```
%cassandra

INSERT INTO cassandra_v3.simple(key,value) VALUES(1, 'first'); 
INSERT INTO cassandra_v3.simple(key,value) VALUES(2, 'second');
INSERT INTO cassandra_v3.simple(key,value) VALUES(3, 'third');

SELECT * FROM cassandra_v3.simple WHERE key={{key=1|2|3}};
```
The result of this paragraph should be:
![image](https://cloud.githubusercontent.com/assets/1532977/15740377/ebb018ea-28b4-11e6-9cde-b3aad50bbaef.png)


* To test the override of `AngularObject` upon dynamic form, create 2 new paragraphs with the following content

First paragraph content:
```
%angular

<form class="form-inline">
  <div class="form-group">
    <label for="keyId">Key: </label>
    <input type="text" class="form-control" id="keyId" placeholder="key id ..." ng-model="key_id"></input>
  </div>
  <button type="submit" class="btn btn-primary" ng-click="z.angularBind('key_id',key_id,'PUT_HERE_SECOND_PARAGRAPH_ID'); z.runParagraph('PUT_HERE_SECOND_PARAGRAPH_ID')"> Bind</button>
</form>
```
Second paragraph content:
```
%cassandra

INSERT INTO cassandra_v3.simple(key,value) VALUES(1, 'first'); 
INSERT INTO cassandra_v3.simple(key,value) VALUES(2, 'second');
INSERT INTO cassandra_v3.simple(key,value) VALUES(3, 'third');

SELECT * FROM cassandra_v3.simple WHERE key={{key_id=1}};
```
Replace `PUT_HERE_SECOND_PARAGRAPH_ID` in the angular code of the first paragraph by the real paragraph id of the second paragraph. Execute the first paragraph, put a value into the input text (3) and click on **Bind**, you should see the following result:

![image](https://cloud.githubusercontent.com/assets/1532977/15740626/e9d9deba-28b5-11e6-9708-f040a8d58ccd.png)


### Screenshots (if appropriate)

### Questions:
* Does the licenses files need update? --> **No**
* Is there breaking changes for older versions? --> **No**
* Does this needs documentation? --> **Yes**

[ZEPPELIN-901]: https://issues.apache.org/jira/browse/ZEPPELIN-901
[Cassandra 3.5]: http://cassandra.apache.org/download/
